packages/common/core-utils/src/assert.ts

```diff
@@ -22,3 +34,112 @@ export function assert(condition: boolean, message: string | number): asserts co
 		);
 	}
 }
+
+/**
+ * Asserts that can be conditionally enabled in debug/development builds but will be optimized out of production builds.
+ *
+ * Disabled by default.
+ *
+ * If the assert must be enforced/checked in production or enabled by default, use {@link assert} instead.
+ *
+ * @param predicate - A pure function that should return true if the condition holds, or a string or object describing the condition that failed.
+ * This function will only be run in some configurations so it should be pure, and only used to detect bugs (when debugAssert are enabled), and must not be relied on to enforce the condition is true: for that use {@link assert}.
+ * @remarks
+ * Remarks optimizing the asserts out of the bundle requires a bundler like webpack which leverages `__PURE__` annotations like https://webpack.js.org/guides/tree-shaking/#mark-a-function-call-as-side-effect-free.
+ *
+ * Exceptions thrown by this function must never be caught in production code as that will result in different behavior when testing and when running optimized builds.
+ * The `predicate` function must be pure (have no side-effects) to ensure that the behavior of code is the same regardless of if the asserts are disabled, enabled or optimized out.
+ *
+ * These asserts are disabled by default, even in debug builds to ensure that by default code will be tested as production runs, with them disabled.
+ * Additionally this ensures that apps that use a bundler which does not remove `__PURE__` will not incur the runtime cost of calling the predicate.
+ * These asserts can be can be enabled by calling `configureDebugAsserts(true)`: see {@link configureDebugAsserts}.
+ *
+ * @privateRemarks
+ * This design was chosen to accomplish two main goals:
+ *
+ * 1. Make it easy to compile debug asserts fully out of production builds.
+ * For webpack this happens by default, avoiding the need for customers to do special configuration.
+ * This is important for both performance and bundle size.
+ * 2. Make it easy to test (both manually and automated) with and without the predicates running.
```

Comment:
Nit (I know intellisense handles this correctly, but technically there needs to be a line break between these per tsdoc spec if you want a line break between the list items)
```suggestion
 *
 * 2. Make it easy to test (both manually and automated) with and without the predicates running.
```